src/context.rs

```diff
@@ -471,18 +474,35 @@ impl PySessionContext {
         name: Option<&str>,
         py: Python,
     ) -> PyResult<PyDataFrame> {
-        // Instantiate pyarrow Table object & convert to batches
-        let table = data.call_method0("to_batches")?;
+        let mut batches = None;
+        let mut schema = None;
 
-        let schema = data.getattr("schema")?;
-        let schema = schema.extract::<PyArrowType<Schema>>()?;
+        if let Ok(stream_reader) = ArrowArrayStreamReader::from_pyarrow_bound(&data) {
+            // Works for any object that implements __arrow_c_stream__ in pycapsule.
+
+            schema = Some(stream_reader.schema().as_ref().to_owned());
+            batches = Some(stream_reader.filter_map(|v| v.ok()).collect());
+        } else if let Ok(array) = RecordBatch::from_pyarrow_bound(&data) {
+            // While this says RecordBatch, it will work for any object that implements
+            // __arrow_c_array__ in pycapsule.
+
+            schema = Some(array.schema().as_ref().to_owned());
+            batches = Some(vec![array]);
+        }
+
+        if batches.is_none() || schema.is_none() {
+            return Err(PyTypeError::new_err(
+                "Expected either a Arrow Array or Arrow Stream in from_arrow_table().",
+            ));
+        }
+
+        let batches = batches.unwrap();
+        let schema = schema.unwrap();
```

Comment:
I'd personally opt for a different control flow that I find slightly clearer, but just a nit:
```suggestion
        let (schema, batches) = if let Ok(stream_reader) = ArrowArrayStreamReader::from_pyarrow_bound(&data) {
            // Works for any object that implements __arrow_c_stream__ in pycapsule.

            schema = Some(stream_reader.schema().as_ref().to_owned());
            batches = Some(stream_reader.filter_map(|v| v.ok()).collect());
            (schema, batches)
        } else if let Ok(array) = RecordBatch::from_pyarrow_bound(&data) {
            // While this says RecordBatch, it will work for any object that implements
            // __arrow_c_array__ in pycapsule.

            schema = Some(array.schema().as_ref().to_owned());
            batches = Some(vec![array]);
            (schema, batches)
        } else {
            return Err(PyTypeError::new_err(
                "Expected either a Arrow Array or Arrow Stream in from_arrow_table().",
            ));        
        }
```